### PR TITLE
Additional information about SPAs

### PR DESCRIPTION
**Pull request by gdestuynder**
including relaxing the "do not use them" since they're used anyway. Instead strongly discourage them and indicate how to use them as safely as possible if you cannot follow that recommendation

### CODE REVIEW

File: docs/guidelines/iam/openid_connect.md [hunk 1]
```diff
@@ -80,19 +81,19 @@ For that reason, it is important that the web application (RP) respects the foll
 
 ### Authorization Code Grant
 
--   **Always** use authorization code grant.
--   **Never** use implicit grants for websites. Authorization code grant ensures that the relying party is getting the access tokens, and that these cannot be intercepted within the user's browser.
+-   **Recommend** using the authorization code grant. This requires a back-end server, reverse-proxy, or similar server-side code that will generate an `httponly` session cookie for the user, which means the access tokens cannot be intercepted in Javascript within the user's browser.
```
Comment (april) [inline docs/guidelines/iam/openid_connect.md]: It's case insensitive, but you may as well use the official spelling of `HttpOnly`. Also JavaScript and "reverse proxy".

File: docs/guidelines/iam/openid_connect.md [hunk 2]
```diff
@@ -67,8 +67,9 @@ For that reason, it is important that the web application (RP) respects the foll
     -   This ensures that access is revoked within *15 minutes* in the event that the user's account is disabled by the OpenID Connect Provider (OP).
     -   This issues a new ID token, with new attributes if they have changed.
     -   This may also renew the ID token expiration time.
-    -   This is generally done with the parameter `prompt=none` while calling the OpenID Connect `authorize` endpoint. See also [specifications](https://openid.net/specs/openid-connect-implicit-1_0.html#RequestParameters).
--   The web application (RP) can **optionally** provide a `logout` URL, which the OpenID Connect Provider (OP) can call to indicate if a user has logged out (so that the web application immediately know when to log the user out as well).
+    -   This is generally done with the parameter `prompt=none` while calling the OpenID Connect `authorize` endpoint. See also [specifications](https://openid.net/specs/openid-connect-implicit-1_0.html#RequestParameters).\
+    - This can also be done in the back-end to avoid a round-trip in the user-agent (user's web browser).
```
Comment (april) [inline docs/guidelines/iam/openid_connect.md]: "user agent".

File: docs/guidelines/iam/openid_connect.md [hunk 3]
```diff
@@ -80,19 +81,19 @@ For that reason, it is important that the web application (RP) respects the foll
 
 ### Authorization Code Grant
 
--   **Always** use authorization code grant.
--   **Never** use implicit grants for websites. Authorization code grant ensures that the relying party is getting the access tokens, and that these cannot be intercepted within the user's browser.
+-   **Recommend** using the authorization code grant. This requires a back-end server, reverse-proxy, or similar server-side code that will generate an `httponly` session cookie for the user, which means the access tokens cannot be intercepted in Javascript within the user's browser.
+-   **Avoid** using implicit grants for websites when possible. Single Page Applications (SPA) use implicit grants and receive tokens back in the user's browser with no server-side component, which means it can be intercepted by malicious javascript running in the user's browser. If using the implicit grant **do not** store the tokens in the browser's storage (do not use localStorage). Instead just keep them in memory. This makes it much harder for attackers using a vulnerability in your website to steal the token.
```
Comment (april) [inline docs/guidelines/iam/openid_connect.md]: What is "it" that can be intercepted? Also JavaScript again.  :)

Maybe also:
Instead, keep them in memory, which much harder for attackers using a vulnerability in your website to steal the token.

File: docs/guidelines/iam/openid_connect.md [hunk 4]
```diff
@@ -80,19 +81,19 @@ For that reason, it is important that the web application (RP) respects the foll
...
 
 #### Additional notes on Implicit grants
-Implicit grants are inherently more dangerous and harder to implement safely. We advocate against their usage entirely.
+Implicit grants are inherently more dangerous and harder to implement safely. We advocate against their usage entirely where possible.
```
Comment (april) [inline docs/guidelines/iam/openid_connect.md]: harder -> difficult would sound cleaner, I think. Also "entirely where possible" sounds a bit weird, either just "entirely" or "whenever possible".

File: docs/guidelines/iam/openid_connect.md [hunk 5]
```diff
@@ -80,19 +81,19 @@ For that reason, it is important that the web application (RP) respects the foll
...
 
-This can become dangerous when the SPA is vulnerable to XSS or CSRF attacks (e.g. attacks where the attacker may retrieve the user's tokens). Many SPAs tend to store the user's `id_token` in the browser LocalStorage as a "proof of authentication" and access tokens which are then used to query otherwise private API endpoints. Unlike browser cookies which can be set to http-only, LocalStorage can be queried in javascript and thus through an XSS vulnerability. 
+This can become dangerous when the SPA is vulnerable to XSS or CSRF attacks (e.g. attacks where the attacker may retrieve the user's tokens). Many SPAs tend to store the user's `id_token` in the browser LocalStorage as a "proof of authentication" and access tokens which are then used to query otherwise private API endpoints. **This is dangerous**. Unlike browser cookies which can be set to http-only, LocalStorage can be queried in javascript and thus through an XSS vulnerability.
```
Comment (april) [inline docs/guidelines/iam/openid_connect.md]: `localStorage` for consistently. HTTP should probably be capitalized as well, same with JavaScript, and `httpOnly`.

File: docs/guidelines/iam/openid_connect.md [hunk 6]
```diff
@@ -80,19 +81,19 @@ For that reason, it is important that the web application (RP) respects the foll
...
 
 Any vulnerability in your SPA may leak the user's tokens (with functional API access) that are used by your SPA. In addition, any **other** SPA that you do not control may also leak the same tokens. These tokens, if allowed for the same APIs, can be then used to compromise your application/APIs.
 
-Authorization code grants prevent this problem by verifying everything on the web-server and preventing the user-agent from accessing this data.
+Authorization code grants prevent this problem by verifying everything on the web-server and preventing the user-agent from accessing this data. Note that for certain websites it may be convenient to perform the authorization code grant at the API level, so that your SPA does not need to manipulate tokens.
```
Comment (april) [inline docs/guidelines/iam/openid_connect.md]: "web server" and "user agent"

File: docs/guidelines/iam/openid_connect.md [hunk 7]
```diff
@@ -102,7 +103,7 @@ This is a defense against [CSRF](https://en.wikipedia.org/wiki/Cross-site_reques
 
 ### Refresh token
 
-Avoid using or storing refresh tokens. This is especially important for relying parties (RP) which are websites (as opposed to mobile apps for example, which may not always have network access). Refresh tokens never expire and thus are very powerful. These are usually not needed for an authentication flow, though they may be needed for specific authorization flows.
+Avoid using or storing refresh tokens. This is especially important for relying parties (RP) which are websites (as opposed to mobile apps for example, which may not always have network access). Refresh tokens never expire and thus are very powerful. These are usually not needed for user authentication flows.
```
Comment (april) [inline docs/guidelines/iam/openid_connect.md]: Can probably remove the parenthesis.

Comment (gdestuynder) [reply docs/guidelines/iam/openid_connect.md]: I like em ! :)

**Pull request merged**